### PR TITLE
tests: adding fedora 31

### PR DESCRIPTION
**Pull request by sergiocazzolato**
The idea is to move it to google backend after the tests are passing
without errors during few days.

### CODE REVIEW

**Comment by bboozzoo** [comment]
Thought I could salvage some stuff from #7193 but once I rebased that PR there was nothing useful left.

**Comment by bboozzoo** [comment]
This should run successfuly now. Running spread locally I get:
```
2019-11-07 15:01:49 Restoring google-unstable:fedora-31-64...                 
2019-11-07 15:01:50 Discarding google-unstable:fedora-31-64 (nov071324-531042)... 
2019-11-07 15:01:51 Successful tasks: 368                                              
2019-11-07 15:01:51 Aborted tasks: 0      
```

File: tests/main/document-portal-activation/task.yaml [hunk 1]
```diff
@@ -41,18 +41,29 @@ restore: |
     rm -rf "${XDG_RUNTIME_DIR:?}/*" "${XDG_RUNTIME_DIR:?}/.[!.]*"
 
 execute: |
+    #shellcheck source=tests/lib/systems.sh
+    . "$TESTSLIB"/systems.sh
+
+    check_stderr() {
+        test -n "$1"
+        if is_cgroupv2; then
+            [[ "$(cat "$1")" == "WARNING: cgroup v2 is not fully supported yet, proceeding with partial confinement" ]]
```
Comment (zyga) [inline tests/main/document-portal-activation/task.yaml]: Why `[[`? Would this work?

```suggestion
            [ "$(cat "$1")" = "WARNING: cgroup v2 is not fully supported yet, proceeding with partial confinement" ]
```

File: tests/main/interfaces-calendar-service/task.yaml [hunk 2]
```diff
@@ -7,7 +7,8 @@ summary: Ensure that the calendar-service interface works
 # https://github.com/snapcore/snapd/pull/7230 is landed
 # ubuntu-19.10: test-snapd-eds is incompatible with eds version shipped with the distro
 # arch-linux: test-snapd-eds is incompatible with eds version shipped with the distro
-systems: [-ubuntu-core-*, -ubuntu-14.04-*, -amazon-*, -centos-*, -opensuse-tumbleweed-*, -ubuntu-19.10-*, -arch-linux-*, -debian-sid-*]
+# fedora-31: test-snapd-eds is incompatible with eds version shipped with the distro
```
Comment (zyga) [inline tests/main/interfaces-calendar-service/task.yaml]: should we change the interface provided by the system to address this? Seeing how all new distros are unsupported makes me wonder if the interface slot needs to be versioned.

File: tests/main/interfaces-contacts-service/task.yaml [hunk 3]
```diff
@@ -6,7 +6,8 @@ summary: Ensure that the contacts-service interface works
 # ubuntu-19.10: test-snapd-eds is incompatible with eds shipped with the distro
 # arch-linux: test-snapd-eds is incompatible with eds version shipped with the distro
 # opensuse-tumbleweed: test-snapd-eds is incompatible with eds version shipped with the distro
-systems: [-ubuntu-core-*, -ubuntu-14.04-*, -amazon-*, -centos-*, -ubuntu-19.10-*, -arch-linux-*, -debian-sid-*, -opensuse-tumbleweed-*]
+# fedora-31: test-snapd-eds is incompatible with eds version shipped with the distro
+systems: [-ubuntu-core-*, -ubuntu-14.04-*, -amazon-*, -centos-*, -ubuntu-19.10-*, -arch-linux-*, -debian-sid-*, -opensuse-tumbleweed-*, -fedora-31-*]
```
Comment (zyga) [inline tests/main/interfaces-contacts-service/task.yaml]: Same question as above.

File: tests/main/refresh-app-awareness/task.yaml [hunk 4]
```diff
@@ -1,4 +1,6 @@
 summary: Ensure that foreground applications block app refresh.
+# fedora-31: uses cgroupv2, which we don't support
+systems: [ -fedora-31-*]
```
Comment (zyga) [inline tests/main/refresh-app-awareness/task.yaml]: We soon will :)

File: tests/smoke/install/task.yaml [hunk 5]
```diff
@@ -1,25 +1,44 @@
...
     echo "Ensure that the snap can be run as root"
     test-snapd-tools.echo hello > stdout.log 2> stderr.log
     MATCH "^hello$" < stdout.log
-    if [ -s stderr.log ]; then
-        echo "stderr.log must be empty but it is not: (run as root)"
-        cat stderr.log
-        exit 1
+    if is_cgroupv2; then
+        # snap-confine issues a warning on a cgroupv2 system
+        [[ "$(cat stderr.log)" == 'WARNING: cgroup v2 is not fully supported yet, proceeding with partial confinement' ]]
```
Comment (zyga) [inline tests/smoke/install/task.yaml]: Same comment as earlier (about `[[` vs `[`)

File: tests/smoke/install/task.yaml [hunk 6]
```diff
@@ -1,25 +1,44 @@
...
     echo "Ensure that the snap can be run as the user"
     su -l -c "test-snapd-tools.echo hello > stdout.log 2> stderr.log" test
     MATCH "^hello$" < stdout.log
-    if [ -s stderr.log ]; then
-        echo "stderr.log must be empty but it is not: (run as user)"
-        cat stderr.log
-        exit 1
+    if is_cgroupv2; then
+        # snap-confine issues a warning on a cgroupv2 system
+        [[ "$(cat stderr.log)" == 'WARNING: cgroup v2 is not fully supported yet, proceeding with partial confinement' ]]
```
Comment (zyga) [inline tests/smoke/install/task.yaml]: Same comment as earlier (about `[[` vs `[`)

File: tests/main/interfaces-calendar-service/task.yaml (hunk 2 above)
Comment (bboozzoo) [reply tests/main/interfaces-calendar-service/task.yaml]: AFAIK there's nothing we can do in snapd. The client library expects the evolution-data-server to register itself under a specific name, which also encodes the API version. The newer versions of eds have a different API version.

Comment (stolowski) [reply tests/main/interfaces-calendar-service/task.yaml]: Nothing to do in this PR (apart from disabling fc31), but I agree we should think about versioning this interface, and maybe filter out the old version of this interface on systems known to require the new one.
I feel we will some more of those kind of API incompatibilities in the future.

File: tests/main/xdg-settings/task.yaml [hunk 7]
```diff
@@ -69,6 +71,17 @@ execute: |
     touch /usr/bin/zenity
     mount --bind /bin/true /usr/bin/zenity
 
+    if is_cgroupv2 ; then
+        # we are expecting the test to fail on a cgroup v2 system
+        if ensure_xdg_settings_output "http://smoke-test-cgroupv2.com" 2> stderr.log ; then
```
Comment (stolowski) [inline tests/main/xdg-settings/task.yaml]: We should probably `rm -f stderr.log` in restore section.

File: tests/smoke/install/task.yaml [hunk 8]
```diff
@@ -1,25 +1,44 @@
 summary: Check that installing and running a snap works
 
+debug: |
+    if test -e stderr.log; then
+        echo "content of stderr.log"
+        cat stderr.log
```
Comment (stolowski) [inline tests/smoke/install/task.yaml]: Same remark about restore & stderr.log.

File: tests/main/xdg-settings/task.yaml (hunk 7 above)
Comment (sergiocazzolato) [reply tests/main/xdg-settings/task.yaml]: @stolowski, thanks for reviewing, it is automatically removed. The test directory is saved before the test execution and restored after the execution, so it is not needed to clean up the test directory.

**Pull request merged**